akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -514,7 +514,29 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
   /**
    * Requests to stop receiving events from a given input port. Cancelling clears any ungrabbed elements from the port.
    */
-  final protected def cancel[T](in: Inlet[T], cause: Throwable): Unit = interpreter.cancel(conn(in), cause)
+  final protected def cancel[T](in: Inlet[T], cause: Throwable): Unit = cancel(conn(in), cause)
+
+  private[this] lazy val _delayedCancellationCallback = getAsyncCallback[(Connection, Throwable)] {
+    case (connection, cause) =>
+      interpreter.log.debug(s"Now cancelling $connection in stage $this")
+      doCancel(connection, cause)
+  }
```

Comment:
Probably not super important since it is not touched often, but this will always be accessed on the graph stage thread so could be a cheaper OptionVal instead of synchronized

Reply:
There's no contention, though, so it will be basically free. Not sure if it makes sense to complicate matters for that.

But there might be another problem that, instead, it is currently only accessed (and likely initialized) from the Scheduler thread. That seems dangerous as well.

Maybe I'll just create a new async callback per cancellation. Since there will be only a small, limited numbers of cancellations per stage that should be fine and similarly cheap.

Reply:
is it? `materializer.scheduleOnce` delegates to `system.scheduler.scheduleOnce` in the end if I read it correctly.
is it even safe to call `getAsyncCallback` from other threads? the AsyncCallback itself is thread-safe, but is `getAsyncCallback`? it's backed by a plan `var callbacksWaitingForInterpreter: List`

On the other hand I see that we use it like that from `TimerGraphStageLogic.scheduleOnce`.

Reply:
my comment was before I saw that Johannes noticed the same thing

Reply:
I missed that, that looks dangerous, `getAsyncCallback` looks at several vars, probably a bug waiting to happen in `scheduleOnce` as well?

Reply:
I created issue https://github.com/akka/akka/issues/27999 for `TimerGraphStageLogic.scheduleOnce` and docs

Reply:
Funny enough the `lazy val` here was only necessary because of the position in `GraphStage`. By moving the definition down below the async callback state variables it could also just be a `val`. But for simplicity I just create a new callback for every call as explained above.